framework/src/wifi_manager/wifi_manager.c

```diff
@@ -237,6 +346,43 @@ void __tizenrt_manual_linkset(const char *msg)
 #endif
 
 
+static void *_reconn_worker(void *arg)
+{
+	_wifimgr_conn_info_msg_s *msg = (_wifimgr_conn_info_msg_s *)arg;
+	wifi_manager_ap_config_s *apinfo = msg->config;
+	wifi_manager_reconnect_config_s *conn_config = msg->conn_config;
+	WIFIMGR_CHECK_RESULT_CLEANUP(_wifimgr_connect_ap(apinfo), "connecting to ap fail\n", NULL, WIFIMGR_FREE_CONNMSG(msg));
+	while (true) {
+		nvdbg("T%d wait signal\n", getpid());
+		WIFIMGR_WAIT_RECONNECT_EVENT;
+		nvdbg("T%d wake up\n", getpid());
+
+		LOCK_RECONN;
+
+		if (g_manager_info.terminate) {		// check terminate
+			UNLOCK_RECONN;
+			nvdbg("T%d exit while loop\n");
+			break;
+		}
+		UNLOCK_RECONN;
+
+		// get next interval
+		uint32_t interval = conn_config->interval;
+
+		// check wait time
+		nvdbg("wait for %d seconds\n", interval);
+		sleep(interval);
+
+		// send do reconnect using wifi_utils
+		nvdbg("send do reconnect using wifi_utils\n");
+		WIFIMGR_CHECK_RESULT_CLEANUP(_wifimgr_connect_ap(apinfo), "connecting to ap fail\n", NULL, WIFIMGR_FREE_CONNMSG(msg));
```

Comment:
Here _wifimgr_connect_ap is executed by a thread. Then, there might be a race condition between this thread and the thread which executes the wifi_manager handlers. Need to be protected.

Reply:
LOCK_WIFIMGR will protect race condition

Reply:
I'm still wondering that LOCK_WIFIMGR protects just the handler thread but not the thread for reconnecting.

Reply:
Sorry I was confused. actually, if wifimanager is in reconnecting_state then other tasks can't request anything to wifi_manager except reconn_worker. so I think calling 'wifi_connect_ap' doesn't need to be protected.

Reply:
Still there might be a possibility unless you move 'WIFIMGR_SET_STATE(WIFIMGR_STA_RECONNECTING);' at line 828 right above pthread_create().

Reply:
That is, before you change the state to WIFIMGR_STA_RECONNECTING, the current state is 'connected'. Then in this state other wifi_manager functions like disconnect can be called.

Reply:
As we discussed, in the state of 'reconnecting', the disconnect api may be called. Then the reconnection thread should be canceled.